lib/awful/rules.lua

```diff
@@ -205,36 +206,102 @@ function rules.apply(c)
+--
+-- To add a new properties, just do:
+--
+--    function awful.rules.extra_properties.my_new_property(c, value)
+--        -- do something
+--    end
+--
+-- @tfield table awful.rules.extra_properties
+rules.extra_properties = {}
+
+--- Extra high priority properties.
+--
+-- Some properties, such as anything related to tags, geometry or focus, will
+-- cause a race condition if set in the main property section. This is why
+-- they have a section for them.
+--
+-- To add a new properties, just do:
+--
+--    function awful.rules.high_priority_properties.my_new_property(c, value)
+--        -- do something
+--    end
+--
+-- @tfield table awful.rules.high_priority_properties
+rules.high_priority_properties = {}
+
+rules.delayed_properties = {
+    focus = true,
+}
 
 --- Apply properties and callbacks to a client.
 -- @client c The client.
 -- @tab props Properties to apply.
 -- @tab[opt] callbacks Callbacks to apply.
 function rules.execute(c, props, callbacks)
-    local handle_later = { focus = true, switchtotag = true }
-    local switchtotag = props.switchtotag
+
+    -- Some properties need to be handled first. For example, many properties
+    -- depend that the client is tagged, this isn't yet the case.
+    for prop, handler in pairs(rules.high_priority_properties) do
+        local value = props[prop]
+
+        if value ~= nil then
+            if type(value) == "function" then
+                value = value(c, props)
+            end
+
+            handler(c, props[prop])
+        end
+
+    end
 
     for property, value in pairs(props) do
+
+
         if property ~= "focus" and type(value) == "function" then
             value = value(c)
         end
-        if property == "screen" then
-            -- Support specifying screens by name ("VGA1")
-            c.screen = screen[value]
-        elseif property == "tag" then
-            local t = value
-            if type(t) == "string" then
-                t = atag.find_by_name(props.screen, t)
-            end
-            c.screen = t.screen
-            c:tags({ t })
-        elseif property == "height" or property == "width" or
-                property == "x" or property == "y" then
-            local geo = c:geometry();
-            geo[property] = value
-            c:geometry(geo);
-        elseif not handle_later[property] then
-            if type(c[property]) == "function" then
+
+        -- Some properties are handled elsewhere
+        if not rules.high_priority_properties[property] and not rules.delayed_properties[property] then
+            if property == "tag" then
+                local t = value
+                if type(t) == "string" then
+                    t = atag.find_by_name(props.screen, t)
+                end
+
+                if type(value) == "function" then
+                    value = value(c, props)
+                end
+
+                if value then
+                    c.screen = value.screen
+                    c:tags{ value }
```

Comment:
This breaks tag-assignment. The new code isn't equivalent to the previous code. Also, the next comment also still applies.

However, the next commit cleans this up again... So the end result is ok, the way to get there is weird.